packages/react-dom/src/client/ReactDOMFiberInput.js

```diff
@@ -314,13 +310,39 @@ function synchronizeDefaultValue(
 ) {
   if (
     // Focused number inputs synchronize on blur. See ChangeEventPlugin.js
-    (type !== 'number' || node.ownerDocument.activeElement !== node) &&
-    node.defaultValue !== '' + value
+    type !== 'number' ||
+    node.ownerDocument.activeElement !== node
   ) {
-    if (value != null) {
-      node.defaultValue = '' + value;
-    } else {
-      node.defaultValue = node._wrapperState.initialValue;
-    }
+    assignProperty(
+      node,
+      'defaultValue',
+      value == null ? node._wrapperState.initialValue : value,
+    );
+  }
+}
+
+function assignProperty(
+  node: InputWithWrapperState,
+  property: string,
+  value: *,
+) {
+  // TODO: This should use DOMProperty.shouldSetAttribute, however the current
+  // behavior is such that invalid attributes do not update the current attribute,
```

Comment:
This sounds like a bug to me. I'd like to always treat them as `null` instead of showing stale values. I'd be fine changing behavior even if that's what we did in 16 because I think it's more likely to cause product bugs.